src/error.rs

```diff
@@ -0,0 +1,210 @@
+    QUIC_ERROR_CERT_NO_CERT
+);
+
+impl From<ErrorCode> for QUIC_ERROR {
+    fn from(value: ErrorCode) -> Self {
+        value as QUIC_ERROR
+    }
+}
+
+/// The display string is the same as the debug string, i.e. the enum string.
+impl core::fmt::Display for ErrorCode {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        core::write!(f, "{:?}", self)
+    }
+}
+
+/// Quic error used in non-ffi code.
+/// Internal representation matches the os platfrom type.
+#[derive(Clone)]
+pub struct Error(pub QUIC_ERROR);
+
+impl Error {
+    /// Create an error from enum.
+    pub fn new(ec: ErrorCode) -> Self {
+        Self(ec as QUIC_ERROR)
+    }
+    /// Convert to error code if possible.
+    pub fn try_as_error_code(&self) -> Result<ErrorCode, &str> {
+        use std::convert::TryFrom;
+        ErrorCode::try_from(self.0 as QUIC_ERROR)
+    }
+
+    /// Convert from raw ffi error type.
+    pub fn from_raw(ec: QUIC_ERROR) -> Self {
+        Self(ec)
+    }
+
+    /// Temp api to use in manually written ffi code which is going to be
+    /// removed and replaced by auto generated ffi code.
+    /// This api will be replaced by from_raw.
+    pub fn from_u32(ec: u32) -> Self {
+        Self(ec as QUIC_ERROR)
+    }
+
+    /// Return Err if the error is considered a failure.
+    /// Ok includes both "no error" and "pending" status codes.
+    pub fn ok(self) -> Result<(), Self> {
+        // on windows it is signed.
+        #[cfg(target_os = "windows")]
+        if self.0 < 0 {
+            Err(self)
+        } else {
+            Ok(())
+        }
+
+        #[cfg(not(target_os = "windows"))]
+        if (self.0 as i32) > 0 {
+            Err(self)
+        } else {
+            Ok(())
+        }
+    }
+}
+
+impl std::error::Error for Error {}
+
+impl From<QUIC_ERROR> for Error {
+    fn from(value: QUIC_ERROR) -> Self {
+        Self(value)
+    }
+}
+
+impl From<ErrorCode> for Error {
+    fn from(value: ErrorCode) -> Self {
+        Self::new(value)
+    }
+}
+
+/// The debug message is in the same format as error in windows crate.
+impl core::fmt::Debug for Error {
+    fn fmt(&self, fmt: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        let mut debug = fmt.debug_struct("Error");
+        let str_code = match self.try_as_error_code() {
+            Ok(c) => Some(c),
+            Err(_) => None,
+        };
+        debug.field("code", &self.0);
+        match str_code {
+            Some(c) => debug.field("message", &c),
+            None => debug.field("message", &"unknown quic error"),
+        };
+        debug.finish()
+    }
+}
+
+/// The display message is in the same format as error in windows crate.
+impl core::fmt::Display for Error {
+    fn fmt(&self, fmt: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        let str_code = match self.try_as_error_code() {
+            Ok(c) => Some(c),
+            Err(_) => None,
+        };
+        match str_code {
```

Comment:
I prefer to print the numeric values as hex please.

Reply:
Done